chandra_aca/centroid_resid.py

```diff
@@ -0,0 +1,332 @@
+                    self.centroid_dt = 0.087
+        else:
+            if len(fetch.data_source.sources()) > 1:
+                raise ValueError("Can't set offsets based on fetch data source if multiple data sources set")
+            fetch_source = fetch.data_source.sources()[0]
+            if fetch_source != 'cxc' and fetch_source != 'maude':
+                raise ValueError("Only maude and cxc fetch data sources are supported for offsets")
+            if is_or:
+                if fetch_source == 'cxc':
+                    if self.att_source == 'ground':
+                        self.centroid_dt = -2.4
+                    if self.att_source == 'obc':
+                        self.centroid_dt = -1.8
+                if fetch_source == 'maude':
+                    if self.att_source == 'ground':
+                        self.centroid_dt = -2.8
+                    if self.att_source == 'obc':
+                        self.centroid_dt = -2.34
+            else:
+                if fetch_source == 'cxc':
+                    if self.att_source == 'ground':
+                        self.centroid_dt = -2.44
+                    if self.att_source == 'obc':
+                        self.centroid_dt = -2.43
+                elif fetch_source == 'maude':
+                    if self.att_source == 'ground':
+                        self.centroid_dt = -2.45
+                    if self.att_source == 'obc':
+                        self.centroid_dt = -2.45
+
+        self.yag_times = self.yag_times + self.centroid_dt
+        self.zag_times = self.zag_times + self.centroid_dt
+        self._dt_applied = True
+
+    def calc_residuals(self):
+        """
+        Calculate residuals based on attitude and ra/dec of star.  Note that the sampling and times
+        of yags may be different from zags so these should be done independently.
+
+        Residuals are available in self.dyags and self.dzags.
+        Predicted values from attitude and star position in self.pred_yags and self.pred_zags
+
+        """
+        if len(self.att_times) < 2:
+            raise ValueError("Cannot attempt to calculate residuals with fewer than 2 attitude samples")
+        eci = Ska.quatutil.radec2eci(self.ra, self.dec)
+        # Transform the 3x3 to get the axes to align to have the dot product make sense
+        d_aca = np.dot(quat_vtransform(self.atts).transpose(0, 2, 1), eci)
+        p_yags = np.arctan2(d_aca[:, 1], d_aca[:, 0]) * R2A
+        p_zags = np.arctan2(d_aca[:, 2], d_aca[:, 0]) * R2A
+        self.pred_yags = interpolate(p_yags, self.att_times, self.yag_times, sorted=True)
+        self.pred_zags = interpolate(p_zags, self.att_times, self.zag_times, sorted=True)
+        self.dyags = self.yags - self.pred_yags
+        self.dzags = self.zags - self.pred_zags
+
+
+    @classmethod
+    def for_slot(cls, obsid=None, start=None, stop=None,
+                 slot=None, att_source='ground', centroid_source='ground'):
```

Comment:
For an ER, the defaults of att_source='ground' and centroid_source='ground' aren't helpful.  Should these be left at 'ground'/'ground' or changed to 'obc'/'obc' for all observations, or should there be some logic in here for ERs?